sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -678,6 +703,42 @@ public Read withPartitionOptions(PartitionOptions partitionOptions) {
               .withTransaction(getTransaction());
       return input.apply(Create.of(getReadOperation())).apply("Execute query", readAll);
     }
+
+    SerializableFunction<Struct, Row> getFormatFn() {
+      return (SerializableFunction<Struct, Row>)
+          input ->
+              Row.withSchema(Schema.builder().addInt64Field("Key").build())
+                  .withFieldValue("Key", 3L)
+                  .build();
+    }
+  }
+
+  public static class ReadRows extends PTransform<PBegin, PCollection<Row>> {
+    Read read;
+    Schema schema;
+
+    public ReadRows(Read read, Schema schema) {
+      super("Read rows");
+      this.read = read;
+      this.schema = schema;
```

Comment:
It would be really great if `SpannerIO.ReadRows` could determine the schema at pipeline construction time so the user doesn't have to specify it. In `SpannerIO.Read#expand` we require the user to specify either a query or a list of columns: https://github.com/apache/beam/blob/2872e37d801b489ecbb2c0d6a2a70430d8ba91e9/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java#L656-L671

In both case we're very close to a schema. We just need to analyze the query and/or get the output types for the projected columns. I looked into it a little bit, but I'm not quite sure the best way to use the spanner client to look up the schema. The only thing I could figure out was to start a read and look at the type of `ResultSet#getCurrentRowAsStruct` which seems less than ideal. 

CC @nielm who's done some work with SpannerIO recently - do you have any suggestions for a way to determine the types of the Structs that SpannerIO.Read will produce?

Reply:
We could also punt on this question and file a jira with a TODO here. I recognize this is a little out of scope for BEAM-10139, BEAM-10140.

Reply:
I'd really like to do it in this PR, but the only thing that comes to mind is to do what you said - perform the read request with client and then read the schema. The obvious disadvantage is that the Spanner query will be executed twice. I researched that limit of 1 row added to the end of query will not improve the performance so this is not the thing to do for huge result sets

Reply:
I can reach out to the Spanner team to see if there's a good way to do this, I'll let you know if I learn anything. For now we can just plan on a jira and a TODO

Reply:
I don't see any good solution here...
When reading an entire table, it could be possible to read the table's schema first, and determine what types the columns are, but this does not work for a query as the query output columns may not correspond to table columns. 

Adding `LIMIT 1` would only work for simple queries, anything with joins, `GROUP BY`, `ORDER BY` will require the majority of the query to be executed before a single row is returned. 

So the only solution I can see is for the caller to specify the row Schema as you do here..

Reply:
It seems like it should be possible to analyze the query and determine the output schema, SqlTransform and JdbcIO both do this

Reply:
Thank you @nielm ! I thought about the LIMIT approach but then I found the same arguments not to do that.

It appears there exist a jdbc client for Spanner: https://cloud.google.com/spanner/docs/jdbc-drivers . I'll try to figure out if I can use it. 

There is ResultSetMetadata in Spanner's REST API which extends json object. https://cloud.google.com/spanner/docs/reference/rest/v1/ResultSetMetadata but at the end of the day it requires at least partially to fetch the data.

But I would leave it for another PR as it supposedly require to move SchemaUtils from io/jdbc to some more general place (extensions/sql?). As I can see Struct type is represented as String as is mentiones here:
```
The Cloud Spanner STRUCT data type is mapped to a SQL VARCHAR data type, accessible through this driver as String types. All other types have appropriate mappings.
```
So it may not be the best option.